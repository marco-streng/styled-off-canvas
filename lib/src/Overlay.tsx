import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Div = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $opacity?: number;
  $duration?: string;
}>(
  ({ $isOpen, $background = "#000", $opacity = 0.3, $duration = "500ms" }) => `
  background: ${$background};
  height: 100%;
  left: 0;
  opacity: ${$isOpen ? $opacity : 0};
  position: fixed;
  top: 0;
  transition: opacity ${$duration} ease-in-out;
  width: 100%;
  z-index: ${$isOpen ? 1 : -1};
`
);

/**
 * Overlay component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Overlay = ({
  className,
  background,
  duration,
  opacity,
  ...props
}: {
  background?: string;
  duration?: string;
  opacity?: number;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  return (
    <Div
      className={className}
      onClick={() => close()}
      $isOpen={isOpen}
      $background={background}
      $duration={duration}
      $opacity={opacity}
      {...props}
    />
  );
};
