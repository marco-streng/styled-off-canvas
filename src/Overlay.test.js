import React from 'react'
import { fireEvent, render } from '@testing-library/react'

import Provider from './Provider'
import Overlay from './Overlay'

describe('Overlay', () => {
  test('default', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen><Overlay /></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('custom color', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen overlayBackground='ccc'><Overlay /></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('custom opacity', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen overlayOpacity={0.5}><Overlay /></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('custom duration', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen overlayDuration='1s'><Overlay /></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('closed', () => {
    const { container } = render(<Provider onClose={() => {}}><Overlay /></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('close on click', () => {
    const cb = jest.fn()
    const { container } = render(<Provider onClose={cb}><Overlay /></Provider>)

    fireEvent.click(container.firstChild)

    expect(cb).toHaveBeenCalled()
  })
})
