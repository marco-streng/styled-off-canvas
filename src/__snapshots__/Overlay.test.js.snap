// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay closed 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 dEPHLF"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom color 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 ippImc"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom duration 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 faOfXh"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom opacity 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 hnRav"
  onClick={[Function]}
  opacity={0.5}
/>
`;

exports[`Overlay default 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 iiubrE"
  onClick={[Function]}
  opacity={0.2}
/>
`;
